File: 38. Count and Say Medium/solution.py
def count_and_say(n: int) -> str:
    if n == 1:
        return "1"
    
    def next_sequence(sequence: str) -> str:
        result = []
        count = 1
        char = sequence[0]

        for i in range(1, len(sequence)):
            if sequence[i] == char:
                count += 1
            else:
                result.append(str(count))
                result.append(char)
                char = sequence[i]
                count = 1
        result.append(str(count))
        result.append(char)

        return "".join(result)
    
    current_sequence = "1"
    for _ in range(1, n):
        current_sequence = next_sequence(current_sequence)

    return current_sequence

print(count_and_say(4))
print(count_and_say(1))

File: 38. Count and Say Medium/solution.ts
function countAndSay(n: number): string {
    if (n === 1) return "1";

    function nextSequence(sequence: string): string {
        let result: string[] = [];
        let count = 1;
        let char = sequence.charAt(0);

        for (let i = 1; i < sequence.length; i++) {
            if (sequence.charAt(i) === char) {
                count++;
            } else {
                result.push(count.toString());
                result.push(char);
                char = sequence.charAt(i);
                count = 1;
            }
        }
        result.push(count.toString());
        result.push(char);

        return result.join("");
    }

    let currentSequence = "1";
    for (let i = 1; i < n; i++) {
        currentSequence = nextSequence(currentSequence);
    }
    return currentSequence;
}

console.log(countAndSay(4))
console.log(countAndSay(1))